### ISSUE TITLE
Practice Paper Part 1 Qn 15

### ISSUE THREAD

**Issue by Brendan8899** [nus-cs2103-AY2425S1/forum#894]
![image](https://github.com/user-attachments/assets/2f840ac4-27fe-4950-98fd-6ba63aa80927)
I refer to question 15. There are actually no correct answers.

The abstract method didnt state that its method is abstract, so removing it will not cause compile error

**Comment by cth06-Github** [comment]
Here's my take for the question: <br>
**Why "`Bar` class is overriding the `process()` method" is true** <br>
`Foo` class has a method called `process()`, and `Bar` inherits class `Foo`, as indicated by the solid arrow with an unshaded arrowhead in the UML diagram. Given that there is a method `process()` in `Bar` as well, which will "clash with" `process()` in `Foo` class, `process()` in `Bar` class should be overridden to make the code work/compile. 

<br>

**Why "Removing the `process()` method from the `Bar` class will not cause a compile error" is true**
As you have mentioned, The abstract method didn't state that its method is abstract, so removing it will not cause compile error. If I may rephrase, it means that `process()` in `Foo` class is not abstract. If I am not wrong, an abstract class need not have an abstract method too. Since `process()` in `Foo` is not abstract, then there is no need to define (implement) the `process()` method in `Bar`.  Thus, removing `process()` from `Bar` class will not affect anything, there won't be compile error. Since `Bar` is a child class of `Foo`, and if `process()` in Bar is removed, when a Bar Object call `process()` (e.g. `barObject.process()` then it is the `Foo's process()` that gets invoked (call the parent method). 

<br>

**Why "Removing the store() method from the Bar class will cause a compile error" is true**
`Goo` is an interface with the `store()` method. Therefore, any class that implements `Goo` must have the method `store()` defined/implemented in their own class, otherwise there is compile error. Since `Bar` implements `Goo`, removing `store()` will result in compile error as it is required of `Bar` to define/implement `store()` as its methods. 

<br>

The question asked for the incorrect statement (not correct statement), so (iv) is the answer as explained in the answer key. Hope this helps.

**Comment by damithc** [comment]
> **Why "`Bar` class is overriding the `process()` method" is true**
> 
> `Foo` class has a method called `process()`, and `Bar` inherits class `Foo`, as indicated by the solid arrow with an unshaded arrowhead in the UML diagram. Given that there is a method `process()` in `Bar` as well, which will "clash with" `process()` in `Foo` class, `process()` in `Bar` class should be overridden to make the code work/compile.

Yup. If the ancestor and the descendant classes have methods with the same signature, it is _overriding_. 

> **Why "Removing the `process()` method from the `Bar` class will not cause a compile error" is true**
> As you have mentioned, The abstract method didn't state that its method is abstract, so removing it will not cause compile error. If I may rephrase, it means that `process()` in `Foo` class is not abstract. If I am not wrong, an abstract class need not have an abstract method too. Since `process()` in `Foo` is not abstract, then there is no need to define (implement) the `process()` method in `Bar`. Thus, removing `process()` from `Bar` class will not affect anything, there won't be compile error. Since `Bar` is a child class of `Foo`, and if `process()` in Bar is removed, when a Bar Object call `process()` (e.g. `barObject.process()` then it is the `Foo's process()` that gets invoked (call the parent method).

Correct. 👍 

> **Why "Removing the store() method from the Bar class will cause a compile error" is true**
> `Goo` is an interface with the `store()` method. Therefore, any class that implements `Goo` must have the method `store()` defined/implemented in their own class, otherwise there is compile error. Since `Bar` implements `Goo`, removing `store()` will result in compile error as it is required of `Bar` to define/implement `store()` as its methods.

Correct. 👍

**Comment by ghos7ie** [comment]
Hi prof @damithc, for "Removing the store() method from the Bar class will cause a compile error"

My understanding is that since we are using Java 17, interfaces do allow for default implementation of methods since Java 8 as seen [here](https://www.techempower.com/blog/2013/03/27/everything-about-java-8/)

![image](https://github.com/user-attachments/assets/03012f46-41ea-412b-91cc-4df280b55a2b)


Does this impact the question or is there a different notation for class diagrams to state that a method in an interface has a default implementation? I think this question can be extended to abstract classes as well, since they also support default implementations now.

**Comment by damithc** [comment]
@ghos7ie You can assume that the methods in the exam questions are not `default`, as the course did not cover a way to indicate default methods.

**Comment by ghos7ie** [comment]
@damithc i see, thanks! will take note of this :)

so it's right to say that for any interface all their methods have not been implemented and for any abstract classes, as long as their methods do not have `{abstract}`, those methods have an implementation?

**Comment by damithc** [comment]
> so it's right to say that for any interface all their methods have not been implemented and for any abstract classes, as long as their methods do not have `{abstract}`, those methods have an implementation?

@ghos7ie Yes, that's correct.
